### PR TITLE
BMqfOfOf: Stop validating self-service certs against trust chain

### PR DESCRIPTION
**Pull request by alan-gds**
https://trello.com/c/BMqfOfOf/542-stop-validating-certificates-against-federation-certificate-chain-when-serving

### CODE REVIEW

File: hub/config/src/main/java/uk/gov/ida/hub/config/application/CertificateService.java [hunk 1]
```diff
@@ -50,6 +55,12 @@ public  Certificate encryptionCertificateFor(String entityId) {
 
     public List<Certificate> signatureVerificationCertificatesFor(String entityId) {
         T config = getConfig(entityId);
+
+        config.getSignatureVerificationCertificates()
+                .stream()
+                .filter(cd -> !certificateValidityChecker.isValid(cd))
+                .forEach(cd -> LOG.warn("Signature verification certificates were requested for entityId '{}' but at least one is invalid", entityId));
+
```
Comment (kerrr) [inline hub/config/src/main/java/uk/gov/ida/hub/config/application/CertificateService.java]: We could call `getSignatureVerificationCertificates().stream()` once then partition on `isValid()`. We could then handle the true case and the false case. Might be nicer?

**Comment by kerrr** [comment]
Looking good. I think we need to test that we are logging in the CertificateService correctly and we'll need an integration test to exercise the this. I can look at the integration test if your having issues with them.

File: hub/config/src/main/java/uk/gov/ida/hub/config/application/CertificateService.java [hunk 2]
```diff
@@ -50,9 +58,16 @@ public  Certificate encryptionCertificateFor(String entityId) {
...
     public List<Certificate> signatureVerificationCertificatesFor(String entityId) {
         T config = getConfig(entityId);
-        return config.getSignatureVerificationCertificates()
+
+        Map<Boolean, List<Certificate>> certsByValidity = config.getSignatureVerificationCertificates()
+                .stream()
+                .collect(partitioningBy(cd -> certificateValidityChecker.isValid(cd)));
+
+        certsByValidity.get(false)
+                .forEach(cd -> LOG.warn("Signature verification certificates were requested for entityId '{}' but at least one is invalid", entityId));
```
Comment (kerrr) [inline hub/config/src/main/java/uk/gov/ida/hub/config/application/CertificateService.java]: I think we could word the message better now. We know how many of the certs are invalid at this stage. We could also make sure we don't log more than one message.
something like:
```
int badCertCount = certsByValidity.get(false).size();
        if (badCertCount > 0){
            LOG.warn("Signature verification certificates were requested for entityId '{}' {} invalid", entityId, badCertCount == 1 ? "one is" : "both are");
        }
```

File: hub/config/src/test/java/uk/gov/ida/hub/config/application/CertificateServiceTest.java [hunk 3]
```diff
@@ -288,4 +324,14 @@ public void getAllCertificatesReturnsAllTransactionAndMatchingServiceCertificate
         assertThat(actualCertificateSet).containsAll(expectedCertificateSet);
     }
 
+    private void checkForExpectedLogWarnings(Iterable<String> expectedMessages) {
```
Comment (alan-gds) [inline hub/config/src/test/java/uk/gov/ida/hub/config/application/CertificateServiceTest.java]: This is slight overkill as all of the tests in fact check for exactly one expected message (or don't call the check), but while writing it I was expecting there'd be some cases where we'd need to handle several.  I think the overengineered version will be more resilient against future changes than a simple .get(0) version would be, though, so having written it, I'd prefer to keep it.

File: hub/config/src/main/java/uk/gov/ida/hub/config/application/CertificateService.java [hunk 4]
```diff
@@ -21,7 +21,7 @@
 
 public class CertificateService <T extends CertificateConfigurable<T>> {
 
-    private static final Logger LOG = LoggerFactory.getLogger(ManagedEntityConfigRepository.class);
+    private static final Logger LOG = LoggerFactory.getLogger(CertificateService.class);
```
Comment (kerrr) [inline hub/config/src/main/java/uk/gov/ida/hub/config/application/CertificateService.java]: 😅👍🏻

File: hub/config/src/test/java/uk/gov/ida/hub/config/application/CertificateServiceTest.java (hunk 3 above)
Comment (kerrr) [reply hub/config/src/test/java/uk/gov/ida/hub/config/application/CertificateServiceTest.java]: Yup, LGTM.

**Pull request merged**